sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala

```diff
@@ -239,18 +239,35 @@ class HDFSMetadataLog[T <: AnyRef : ClassTag](sparkSession: SparkSession, path:
       .reverse
   }
 
+  private var lastPurgedBatchId: Long = -1L
+
   /**
    * Removes all the log entry earlier than thresholdBatchId (exclusive).
    */
   override def purge(thresholdBatchId: Long): Unit = {
-    val batchIds = fileManager.list(metadataPath, batchFilesFilter)
-      .map(f => pathToBatchId(f.getPath))
-
-    for (batchId <- batchIds if batchId < thresholdBatchId) {
-      val path = batchIdToPath(batchId)
-      fileManager.delete(path)
-      logTrace(s"Removed metadata log file: $path")
+    val possibleTargetBatchIds = (lastPurgedBatchId + 1 until thresholdBatchId)
+    if (possibleTargetBatchIds.length <= 3) {
+      // avoid using list if we only need to purge at most 3 elements
+      possibleTargetBatchIds.foreach { batchId =>
+        val path = batchIdToPath(batchId)
+        if (fileManager.exists(path)) {
```

Comment:
This confused me. Why not blindly delete the path as the original way does?

Reply:
Original way is not blindly deleting. It does list to enumerate the paths and deletes them. Right?

Reply:
I see, although seems blindly deleting is also fine since it doesn't throw the exception if the file doesn't exist and reduces one more filesystem operation.

Reply:
Yeah that also makes sense. I'm not sure about how much the cost would be saved though. Let me play with this a bit.

Reply:
Sure, please. I'm fine either way unless there's a noticeable perf difference.

Reply:
(Just wanted to mention; the case what I'm considering is when the file doesn't exist - then the case would be changed to exist vs delete. I'm going to evaluate this because I don't know the cost comparison between exist and delete on non-exist - if the cost difference is significant and exist is faster, it's going to be some sort of probability/heuristic. If not, we should simply try calling delete.)

Reply:
I thought about this a bit more, and I admit I was wrong about the assumption; log file may not be existed. It could be for some batches after restarting from checkpoint, but once the query runs more than the threshold (100 batches by default), the log file "should" exist.

Thanks for raising this. Will make a change and run simple test before pushing.

Reply:
Ahh... there's a roadblock on the details - FileSystemBasedCheckpointFileManager.delete logs INFO message when the file doesn't exist. FileContextBasedCheckpointFileManager.delete doesn't log any message though.

Probably OK to remove INFO message on FileSystemBasedCheckpointFileManager.delete as FileSystemBasedCheckpointFileManager is used as a fail-back of FileContextBasedCheckpointFileManager and FileContextBasedCheckpointFileManager covers the major usage.

Reply:
delete() is a no-op if the file doesn't exist; it always has to do the check to see if its there (and whether a file/dir), so wrapping it with another check is superfluous. Azure abfs will suffer here, HDFS less, but still two RPCs with one block for NN read access, one for NN r/w. I don't know anything about HDFS NN locking so can't comment there

Reply:
Looks like adding the check doesn't contribute noticeable latency;

before the change

```
grep "walCommit took" driver_stderr.log | awk '{print $7}' | head -n 1100 | tail -n 1000 | datamash max 1 min 1 mean 1 median 1 perc:90 1 perc:95 1 perc:99 1
1874	152	242.131	220.5	306	342	654.84
```

after the change

```
grep "walCommit took" driver_stderr.log | awk '{print $7}' | head -n 1100 | tail -n 1000 | datamash max 1 min 1 mean 1 median 1 perc:90 1 perc:95 1 perc:99 1
1882	156	252.162	231.5	311	357.2	559.16
```

but I agree the check is unnecessary. Will remove it.

Reply:
Addressed via 03801f7

Reply:
thanks for the investigation